bmcs/src/main/scala/akka/stream/alpakka/oracle/bmcs/impl/BmcsStream.scala

```diff
@@ -0,0 +1,302 @@
+   * @param range      optional range of bytes.
+   * @return source built from the object in storage.
+   */
+  def download(bucket: String, objectName: String, range: Option[ByteRange] = None): Source[ByteString, NotUsed] = {
+    import mat.executionContext
+    Source
+      .fromFuture(request(bucket, objectName, range).flatMap(entityForSuccess).map(_.dataBytes))
+      .flatMapConcat(identity)
+  }
+
+  /**
+   * Returns a future httpResponse to download an object.
+   *
+   * @param objectName  the object to download.
+   * @param bucket      the bucket
+   * @param rangeOption optional range
+   * @return a Future of HttpResponse that download the object from bmcs.
+   */
+  def request(bucket: String, objectName: String, rangeOption: Option[ByteRange] = None): Future[HttpResponse] = {
+    val downloadRequest: HttpRequest = getDownloadRequest(bucket, objectName)
+    signAndGet(rangeOption match {
+      case Some(range) => downloadRequest.withHeaders(headers.Range(range))
+      case _ => downloadRequest
+    })
+  }
+
+  /**
+   * Uploads a stream of ByteStrings to a specified location as a multipart upload.
+   */
+  def multipartUpload(bucket: String,
+                      objectName: String,
+                      chunkSize: Int = MinChunkSize,
+                      chunkingParallelism: Int = 4): Sink[ByteString, Future[CompleteMultipartUploadResult]] =
+    chunkAndRequest(bucket, objectName, chunkSize)(chunkingParallelism)
+      .toMat(completionSink(bucket, objectName))(Keep.right)
+
+  /**
+   * Initiates the multipart upload.
+   * https://docs.us-phoenix-1.oraclecloud.com/api/#/en/objectstorage/20160918/MultipartUpload/CreateMultipartUpload
+   *
+   * @param objectName the object that's being uploaded as multipart.
+   * @param bucket     the bucket to upload to
+   * @return The uploadId wrapped in [MultipartUpload]
+   */
+  private def initiateMultipartUpload(bucket: String, objectName: String): Future[MultipartUpload] = {
+    import mat.executionContext
+    Marshal(CreateMultipartUploadDetails(objectName))
+      .to[RequestEntity]
+      .flatMap(
+        entity =>
+          signAndGetAs[MultipartUpload](
+            initiateMultipartUploadRequest(bucket)
+              .withEntity(entity)
+        )
+      )
+  }
+
+  private def chunkAndRequest(bucket: String, objectName: String, chunkSize: Int = MinChunkSize)(
+      parallelism: Int = 4
+  ): Flow[ByteString, PartUploadResponse, NotUsed] = {
+
+    // Multipart upload requests (except for the completion api) are created here.
+    //  The initial upload request gets executed within this function as well.
+    //  The individual upload part requests are created.
+    val requestFlow: Flow[ByteString, (HttpRequest, (MultipartUpload, Int)), NotUsed] =
+      createMultipartRequests(bucket, objectName, chunkSize, parallelism)
+
+    // The individual upload part requests are processed here
+    requestFlow.via(Http().superPool[(MultipartUpload, Int)]()).map {
```

Comment:
Is creating a new superpool for every `chunkAndRequest` call really the right thing to do?

Reply:
I am not really sure. In theory, we can use the same pool for all the requests for all the uploads and downloads to the same region.. ?

Reply:
How does the AWS S3 connector handle this? I think that has seen the most work/thought go into it so following along how that works is probably a good idea.